core/fedavgoptimizer.py

```diff
@@ -0,0 +1,144 @@
+from core.utils.event_types import ActionableEventTypes, RawEventTypes, callback_handler
+from core.utils.dmljob import deserialize_job
+from core.utils.dmlresult import DMLResult
+
+class FederatedAveragingOptimizer(object):
+	"""
+	Federated Averaging Optimizer
+
+	Implements the Federated Averaging algorithm (with some tweaks)
+	and acts as a "brain" for the Unix Service
+
+	"""
+	def __init__(self, content):
+		"""
+		Kwargs is a dictionary containing all the info needed to initialize the Optimizer.
+		This includes any number of properties such as thresholds, etc.
+		"""
+		kwargs = content.get('optimizer_params')
+		serialized_job = content.get('serialized_job')
+		# The Optimizer knows its job
+		self.job = deserialize_job(serialized_job)
+		self.listen_iterations = kwargs.get('listen_iterations', 0)
+		self.listen_bound = kwargs.get('listen_bound', 1)
+		self.LEVEL_1_CALLBACKS = {
+			RawEventTypes.JOB_DONE.value: self.handle_job_done,
+			RawEventTypes.NEW_INFO.value: self.handle_new_info,
+			RawEventTypes.NOTHING.value: self.do_nothing,
+		}
+		self.LEVEL_2_JOB_CALLBACKS = {
+			RawEventTypes.JOB_TRAIN.value: self._done_training,
+			RawEventTypes.JOB_INIT.value: self._done_initializing,
+			RawEventTypes.JOB_AVG.value: self._average_iter,
+			RawEventTypes.NOTHING.value: self.do_nothing,
+		}
+		self.LEVEL_2_INFO_CALLBACKS = {
+			RawEventTypes.NEW_WEIGHTS.value: self._handle_new_weights,
+			RawEventTypes.TERMINATE.value: self._done_everything,
+			RawEventTypes.NOTHING.value: self.do_nothing,
+		}
+
+	def handle_job_done(self, result):
+		"""
+		Level 1 Callback to handle the payload of a DMLResult
+		"""
+		# assert payload is a DMLResult
+		assert isinstance(result, DMLResult)
+		# assume success for now
+		# get the DMLJob from the DMLResult
```

Comment:
This is misleading because the DMLJob is not retrieved from the DMLResult in this method. Consider removing it.

Reply:
done